### PR TITLE
Minor code fixes

### PR DESCRIPTION
**Pull request by jarmonik**
- Removed 'register' keyword.
- fixed link warnings.
- Replaced prt() function with '&' macro.
- Set default standard to C++17

### CODE REVIEW

File: OVP/D3D9Client/D3D9Util.h [hunk 1]
```diff
@@ -71,8 +71,13 @@ const char *_PTR(const void *p);
 
 // helper function to get address of a temporary
 // NB: use with caution
-template<typename T>
-T* ptr(T&& x) { return &x; }
+
+
+// Required only with c++20 without /permissive flag
```
Comment (dimitry-ishenko) [inline OVP/D3D9Client/D3D9Util.h:76]: This is not related to C++20. Taking address for a temporary is invalid C++ code. It fails with GCC and Clang even with C++11 standard: https://godbolt.org/z/e1jjobY9G

MSVC simply allowed this code up until C++20. So if you want portability and being able to use other compilers, this should be reverted.

File: OVP/D3D9Client/D3D9Util.h [hunk 2]
```diff
@@ -71,8 +71,13 @@ const char *_PTR(const void *p);
...
+// template<typename T>
+// T* ptr(T&& x) { return &x; }
+
+#define ptr &	// use for faster code
```
Comment (dimitry-ishenko) [inline OVP/D3D9Client/D3D9Util.h:80]: It's not faster. With /O2 flag MSVC produces identical code for both versions: 

https://godbolt.org/z/jnqvMq4G3

Except this is a valid code, as long as you use the address within the same expression.

**Comment by jarmonik** [comment]
Ok, I change it back. But this is just an other example of language falling apart. Need a hack to solve a very basic issue that's been working fine before. Not my fault.

**Comment by dimitry-ishenko** [comment]
@jarmonik I understand how you feel. 😏  The problem is not really the language, but rather Microsoft allowing and promoting non-standard code for many years. People latched on and got used to it.

**Comment by schnepe2** [comment]
Am I the only one that gets VERY scared when one uses this?
```cpp
#define ptr &
```
It was defined as a macro, so its use would be clear (as you "invoke" it like a function call like `auto foo = ptr(bar);`)
but now every occurance of `ptr` will be replaced by `&`, which is silly - or even dangerous!
Imagine I would like to write code like this:
`bool uptrend = true;` ...
Sure , this example would lead to a syntax error, but imagine this:
```cpp
bool u = false; // some flag with a good name 'u'
bool end = false;; // flag indicatind some end-condition
bool foo = uptrend; // store current uptrend value for later use
```
this would give valid code
```cpp
bool u = false; // some flag with a good name 'u'
bool end = false; // flag indicatind some end-condition
bool foo = u&end; // !!!!
```
This should never be accepted. By no reason!

**Comment by jarmonik** [comment]
I'll suspend my current work in progress and start restoring that fine function immediately.

**Comment by schnepe2** [comment]
What would be the **right way** to adress this?
Change the signture(s) of the method(s) that are using this?
Like change
```cpp
void vObject::RenderAxisVector(D3D9Pad *pSkp, const D3DXCOLOR *pColor, /*...*/);
void vObject::RenderAxisLabel(D3D9Pad *pSkp, const D3DXCOLOR *clr, /*...*/);
```
to use const references?
```cpp
void vObject::RenderAxisVector(D3D9Pad *pSkp, const D3DXCOLOR &Color, /*...*/);
void vObject::RenderAxisLabel(D3D9Pad *pSkp, const D3DXCOLOR &clr, /*...*/);
```

**Comment by jarmonik** [comment]
I have thought about that. But there are a lot of optional parameters and passing a NULL pointer is a common way to indicate (no data available). Which doesn't work with reference. I have googled about that many times. So, using a reference would require more complex solution to a very simple problem. I looked at the ptr() function and didn't quite follow what was it doing exactly. I don't like relaying on a hack solution but I guess there's no other choice. So, I was planning the use the ptr() anyway, a lesser bad.

Do you have any info about what's going on with the unions ?  I heard that they are obsolete since C++11

**Comment by schnepe2** [comment]
Yeah, I also think the `ptr()` macro is currently the "least-worse" of our options.
Maybe there is a way to limit the exception to the [-fpermissive] rule to specific code-units...
Some `#pragma(...)` on some few files.

> Do you have any info about what's going on with the unions ? I heard that they are obsolete since C++11

As far as I remember the changes will not break any existing code since they only **relax** current union rules.

**Comment by jarmonik** [comment]
Thanks about the info. Yes, I tried the [-permissive] compile flag and it works on MSVC but GCC won't support it. Since some of us are aiming on multi-platform, we must set coding standards according to the worst compiler on markets.

**Comment by jarmonik** [comment]
There is plenty of union discussion here: https://github.com/orbitersim/orbiter/pull/368

**Comment by schnepe2** [comment]
Regarding @dimitry-ishenko s comment further up, we could introduce some "common-often-used" VECTORS and use them.
And for those few places we need special VECTORS we still can use temporaries:
Here's my proof-of-concept (based on comment further up)

https://godbolt.org/z/hsarT8588

**Comment by dimitry-ishenko** [comment]
> I have thought about that. But there are a lot of optional parameters and passing a NULL pointer is a common way to indicate (no data available).

@jarmonik C++17 introduced [std::optional](https://en.cppreference.com/w/cpp/utility/optional) for these situations, so you could do this:

```c++
#include <iostream>
#include <optional>

struct vector { double x, y, z; };

auto foo(const std::optional<vector>& v)
{
    if (v == std::nullopt)
        std::cout << "no vector" << std::endl;
    else std::cout << v->x << ' ' << v->y << ' ' << v->z << std::endl;
}

int main()
{
    foo({ });
    foo(vector{5, 6, 7});
    return 0;
}
```

> Do you have any info about what's going on with the unions ? I heard that they are obsolete since C++11

There are not obsolete per se. It's just that people have been abusing them for years. Many, like you, without even realizing it. 😺

**Comment by jarmonik** [comment]
> Just another idea... Would something like this "wrapper template" be a way to go? The changes in the "original" code would be mainly deleting of the '&'...
> 

Maybe a working solution but it looks over complicated due to wrapper class.

**Comment by dimitry-ishenko** [comment]
I thought of different solutions, but in the end came up with the `ptr()` as the simplest solution. It's also expressive in that it tells you that you are extracting a pointer from an expression.

**Comment by schnepe2** [comment]
Fine with me 👍 If the other compilers do like it (when being "pushed to accept it" 😁)
I didn't mean to distract you two too much from the original PRs intention.

**Comment by jarmonik** [comment]
something like this could work too:  But wouldn't make much difference to std::optional<> approach 

```
#include <cmath>

struct vector3 {
	float x, y, z;
	static vector3 null;

        inline bool notNull() { return x != null.x; }
	inline bool operator!= (const vector3& f) const
	{
		return x != f.x || y != f.y || z != f.z;
	}
};

void func(double a, vector3& v)
{
	if (v != vector3::null) {
		//use v
	}
        if (v.notNull()) {
		//use v
	}
}

void testfunc()
{
	// glabally initialize null
	vector3::null = { nanf("0"), nanf("0"), nanf("0") };

	func(1.0, vector3::null);
}
```

**Comment by dimitry-ishenko** [comment]
> something like this could work too

@jarmonik  hehe, it doesn't actually... https://godbolt.org/z/P5cGrbbrs

**Comment by jarmonik** [comment]
Yeah, it looks like it fails to compare them. It works if comparing the bit-fields instead.

**Comment by dimitry-ishenko** [comment]
However, we **can** use your approach if you don't want to use std::optional. But, I would recommend using templates instead of member variables to avoid code duplication. Eg, something like this:

https://godbolt.org/z/xxYsrdTjz

**Pull request merged**